### ISSUE TITLE
SSH channel not release after connection failure

### ISSUE THREAD

**Issue by andre-merzky** [radical-cybertools/saga-python#451]
From Jeremy Cohen:

Dear Mark and Andre,

I'm having an issue when connecting with SAGA-Python to cloud resources that have just started up. This might well be me doing something incorrect but I wonder if you have any thoughts or can offer any advice on this problem.

The issue is that when I start a cloud instance, I need to wait for it to fully boot up and be accessible via the network. libcloud reports that the instance is running as soon as its state in OpenStack switches to "running" and it is assigned a public IP address. However, at this point, the machine is not normally accessible because it hasn't yet booted fully.

I could just put in an arbitrary sleep of, say, 20 seconds but I'd rather have a more reliable way of checking that the resource is available. I tried using some code to create an instance of the Directory class pointing to the root directory on the remote node. I then try to call list() on this object. My understanding is that instantiating the Directory object triggers the SSH connection to the remote node. I was hoping that if the node is not accessible, some sort of error would be thrown and I can then wait for a few seconds before retrying.

I have some code similar to the following:

```
        retries = 0
        node_running = False
        while (retries < 3) and not node_running:
            try:
                directory = Directory('sftp://%s/' % (node_ip), session=self.session)
                directory.list()
                node_running = True
            except NoSuccess as e:
                LOG.debug('Unable to list base directory for node %s. '
                          'Assuming this node is not accessible yet.'
                           % node_ip)
            except Exception as e2:
                LOG.debug('Unable to list base directory for node %s. '
                          'Assuming this node is not accessible yet: %s'
                           % (node_ip, str(e2)))
            LOG.debug('Waiting 10 seconds before retrying node access...')
            retries+=1
            time.sleep(10)
```

What actually happens when this runs is that the remote node throws an error:
NoSuccess: unexpected EOF (ssh: connect to host x.x.x.x port 22: Network is unreachable
…but control is not returned to my code for quite some time. At first I just thought I wasn't able to catch the exception, hence my second general "except Exception" clause to try and see if I was catching the wrong type of exception. However, after waiting for a long time, control does return to my code and the exception is caught.

I wonder if you've had any experience with this? I can take a look at pty_process.py and see if I can understand what's going on but I wondered if this was something you'd encountered at all? I suspect this may be an SSH issue.

The relevant part of the error output is shown below.

Many thanks,

Jeremy

```
2015:08:19 16:06:32 57226  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   28] ( printf 'HELLO_%d_SAGA\n' 3\n)
2015:08:19 16:06:32 57226  MainThread   saga.ShellDirectory   : [DEBUG   ] read : [    4] [   70] (ssh: connect to host x.x.x.x port 22: Network is unreachable\n)
2015:08:19 16:06:32 57226  MainThread   saga.ShellDirectory   : [DEBUG   ] read : MacOS EOF
08-19 16:06 root         DEBUG    Traceback (most recent call last):
  File "/.../saga-python/src/saga/utils/pty_process.py", line 787, in find
    data += self.read (timeout=_POLLDELAY)
  File "/.../saga-python/src/saga/utils/pty_process.py", line 676, in read
    raise e
NoSuccess: unexpected EOF (ssh: connect to host x.x.x.x port 22: Network is unreachable
) (/.../saga-python/src/saga/utils/pty_process.py +606 (read)  :  raise se.NoSuccess ("unexpected EOF (%s)" % self.tail))

08-19 16:06 root         DEBUG    Traceback (most recent call last):
  File "/.../saga-python/src/saga/utils/pty_shell_factory.py", line 302, in _initialize_pty
    n, match = pty_shell.find (prompt_patterns, delay)
  File "/.../saga-python/src/saga/utils/pty_process.py", line 790, in find
    raise ptye.translate_exception (e, "(%s)" % data)
NoSuccess: unexpected EOF (ssh: connect to host x.x.x.x port 22: Network is unreachable
) (/.../saga-python/src/saga/utils/pty_process.py +606 (read)  :  raise se.NoSuccess ("unexpected EOF (%s)" % self.tail))

2015:08:19 16:06:32 57226  MainThread   saga.ShellDirectory   : [DEBUG   ] PTYProcess del  <saga.utils.pty_process.PTYProcess object at 0x10241eb50>
2015:08:19 16:06:32 57226  MainThread   saga.ShellDirectory   : [DEBUG   ] PTYShell del  <saga.utils.pty_shell.PTYShell object at 0x10241e8d0>
08-19 16:16 deployer.openstack_ec2_deployer DEBUG    Unable to list base directory for node x.x.x.x. Assuming this node is not accessible yet: stop waiting on object lease
08-19 16:16 deployer.openstack_ec2_deployer DEBUG    Waiting 10 seconds before retrying node access…
```

**Comment by andre-merzky** [comment]
Hi Jeremy,

"stop waiting on object lease" indicates that your code is not waiting
for the remote host, but rather for a SAGA level ssh handle.  I'll
open a ticket for this one, that is a bug.

A couple of questions: is it already the first connection attempt
which takes so long to recover, or does that only come up after a
while?  Does the same code succeed when you run it while the host is
up already?

Thanks, Andre.

**Comment by jcohen02** [comment]
To answer your questions:

1) Yes, this error occurs on the first connection attempt. However, after some further testing, I can now see that the code shown in my original question with the loop making three retries does actually work, its just that after getting the "Network is unreachable" error, the code hangs for approximately 10 mins! It then continues with the next iteration of the loop where it can successfully connect to the resource.

I also observed a similar issue later in my code where I'm defining an instance of a saga.filesystem.File object. In this case the machine is running but I had forgotten to set the session object when creating the file so the connection fails with "Too many authentication failures" - again, I see a similar NoSuccess error, followed by:

```
2015:08:19 19:21:08 58164  MainThread   saga.ShellFile        : [DEBUG   ] PTYProcess del  <saga.utils.pty_process.PTYProcess object at 0x111078210>
2015:08:19 19:21:08 58164  MainThread   saga.ShellFile        : [DEBUG   ] PTYShell del  <saga.utils.pty_shell.PTYShell object at 0x1112a0b50>
```

After which the code hangs, presumably for another 10 mins.

2) After adding a delay prior to hitting the code that attempts the SSH connection, so that I can be sure that the cloud instance is operational, the connection proceeds without any problem and the code works as expected.

FYI, here's the full debug output from the point where the connection is initially attempted to the point where control is returned to my code after the long delay:

```
08-19 19:10 deployer.openstack_ec2_deployer DEBUG    About to start <1> resources of type <Large Instance> based on image...
2015:08:19 19:10:58 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] PTYShell init <saga.utils.pty_shell.PTYShell object at 0x1111bf910>
2015:08:19 19:10:58 58164  MainThread   saga.ShellDirectory   : [INFO    ] PTY prompt pattern: [\$#%>\]]\s*$
2015:08:19 19:10:58 58164  MainThread   saga.ShellDirectory   : [INFO    ] ssh copy  mode set to 'sftp'
2015:08:19 19:10:58 58164  MainThread   saga.ShellDirectory   : [INFO    ] ssh share mode set to 'auto'
2015:08:19 19:10:58 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] open master pty for [ssh] [user@x.x.x.x] user: /usr/bin/env TERM=vt100  "/usr/bin/ssh" -t -o IdentityFile=<my_key_file>  -o ControlMaster=auto -o ControlPath=/tmp/saga_ssh_user_%h_%p.remote_user.ctrl -o TCPKeepAlive=no  -o ServerAliveInterval=10 -o ServerAliveCountMax=20 user@x.x.x.x'
2015:08:19 19:10:58 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] PTYProcess init <saga.utils.pty_process.PTYProcess object at 0x1111bfbd0>
2015:08:19 19:10:58 58164  MainThread   saga.ShellDirectory   : [INFO    ] running: /usr/bin/env TERM=vt100 /usr/bin/ssh -t -o IdentityFile=<my_key_file> -o ControlMaster=auto -o ControlPath=/tmp/saga_ssh_user_%h_%p.remote_user.ctrl -o TCPKeepAlive=no -o ServerAliveInterval=10 -o ServerAliveCountMax=20 user@x.x.x.x
2015:08:19 19:10:58 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   33] ( export PS1='$' ; set prompt='$'\n)
2015:08:19 19:10:59 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   51] ( export PS1='$' > /dev/null 2>&1 || set prompt='$'\n)
2015:08:19 19:10:59 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   28] ( printf 'HELLO_%d_SAGA\n' 1\n)
2015:08:19 19:11:00 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   51] ( export PS1='$' > /dev/null 2>&1 || set prompt='$'\n)
2015:08:19 19:11:00 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   28] ( printf 'HELLO_%d_SAGA\n' 2\n)
2015:08:19 19:11:01 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   51] ( export PS1='$' > /dev/null 2>&1 || set prompt='$'\n)
2015:08:19 19:11:01 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   28] ( printf 'HELLO_%d_SAGA\n' 3\n)
2015:08:19 19:11:02 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   51] ( export PS1='$' > /dev/null 2>&1 || set prompt='$'\n)
2015:08:19 19:11:02 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   28] ( printf 'HELLO_%d_SAGA\n' 4\n)
2015:08:19 19:11:03 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   51] ( export PS1='$' > /dev/null 2>&1 || set prompt='$'\n)
2015:08:19 19:11:03 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   28] ( printf 'HELLO_%d_SAGA\n' 5\n)
2015:08:19 19:11:04 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   51] ( export PS1='$' > /dev/null 2>&1 || set prompt='$'\n)
2015:08:19 19:11:04 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] write: [    4] [   28] ( printf 'HELLO_%d_SAGA\n' 6\n)
2015:08:19 19:11:04 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] read : [    4] [   70] (ssh: connect to host x.x.x.x port 22: Network is unreachable\n)
2015:08:19 19:11:04 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] read : MacOS EOF
08-19 19:11 root         DEBUG    Traceback (most recent call last):
  File "/.../saga-python/src/saga/utils/pty_process.py", line 787, in find
    data += self.read (timeout=_POLLDELAY)
  File "/.../saga-python/src/saga/utils/pty_process.py", line 676, in read
    raise e
NoSuccess: unexpected EOF (ssh: connect to host x.x.x.x port 22: Network is unreachable
) (/.../saga-python/src/saga/utils/pty_process.py +606 (read)  :  raise se.NoSuccess ("unexpected EOF (%s)" % self.tail))

08-19 19:11 root         DEBUG    Traceback (most recent call last):
  File "/.../saga-python/src/saga/utils/pty_shell_factory.py", line 302, in _initialize_pty
    n, match = pty_shell.find (prompt_patterns, delay)
  File "/.../saga-python/src/saga/utils/pty_process.py", line 790, in find
    raise ptye.translate_exception (e, "(%s)" % data)
NoSuccess: unexpected EOF (ssh: connect to host x.x.x.x port 22: Network is unreachable
) (/.../saga-python/src/saga/utils/pty_process.py +606 (read)  :  raise se.NoSuccess ("unexpected EOF (%s)" % self.tail))

2015:08:19 19:11:04 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] PTYProcess del  <saga.utils.pty_process.PTYProcess object at 0x1111bfbd0>
2015:08:19 19:11:04 58164  MainThread   saga.ShellDirectory   : [DEBUG   ] PTYShell del  <saga.utils.pty_shell.PTYShell object at 0x1111bf910>

```
--------- HANGS FOR APPROX 10 MINS !! ------
```
08-19 19:21 deployer.openstack_ec2_deployer DEBUG    Unable to list base directory for node x.x.x.x. Assuming this node is not accessible yet: stop waiting on object lease
08-19 19:21 deployer.openstack_ec2_deployer DEBUG    Waiting 10 seconds before retrying node access…
```

Let me know if I can provide further info.
Jeremy

**Comment by jcohen02** [comment]
Some further info…the issue can be reproduced using the following code:

```
import os
os.environ['SAGA_VERBOSE'] = 'DEBUG'

import saga
from saga.filesystem import Directory
from saga.exceptions import NoSuccess

if __name__ == '__main__':

    ctx = saga.Context("ssh")
    ctx.user_id = "<user_id>"
    ctx.user_key = "<key_file>"
    session = saga.Session(default=False)
    session.add_context(ctx)
    
    node_ip = 'x.x.x.x'
    
    try:
        directory = Directory('sftp://%s/' % node_ip, session=session)
        dir_list = directory.list()
    except NoSuccess as e:
        print('NoSuccess: Unable to list base directory for node <%s>: %s ' 
                  % (node_ip, str(e)))
    except Exception as e2:
        print('Exception: Unable to list base directory for node <%s>: %s ' 
                  % (node_ip, str(e2)))
```

Setting node_ip to an IP where the SSH port is accessible but connection fails for some reason - e.g. SSH server is not running on port 22, connection refused,  connection dropped due to invalid key, etc. reproduces the issue.

Looks like the issue is related to the shell lease not being released in the event of such an error. There is an attempt to obtain a lease from the pool and the code ends up stuck in the timeout loop in [lease_manager.py](https://github.com/radical-cybertools/radical.utils/blob/3387b95b7af605e52bf11059b42db738ca6b15b6/src/radical/utils/lease_manager.py#L305)

**Comment by andre-merzky** [comment]
Thanks, that is very useful!  -- A.

**Comment by andre-merzky** [comment]
Dear Jeremy, please let us know if this fix in radical.utils (devel branch) fixes your problem.

Thanks! Andre.

**Comment by jcohen02** [comment]
Andre,

Looks like this fixes the immediate problem I was experiencing. Many thanks.
When there is no connectivity to a remote node, the connection failure is no longer getting stuck in the long timeout. 

One thing that would be really nice to have (and it may be in there already but I haven't spotted it) is the ability to set a timeout on the time spent attempting to connect to a remote node. Perhaps this could/can be done on the session or security context?

Thanks, Jeremy.

**Comment by andre-merzky** [comment]
Point taken on the timeout...  At the moment it is fixed to 100 prompt-read-attempts: https://github.com/radical-cybertools/saga-python/blob/devel/src/saga/utils/pty_shell_factory.py#L284
but that could indeed be made configurable.  Using the context for that is a neat proposal -- but we are limited wrt. API changes by the SAGA API spec.  What we do  support though is configuration via a config file ($HOME/saga.cfg) and via env variables.  If either of those is agreeable for you,  I would not mind implementing a proper timeout, so please let us know.

Best, Andre.

**Comment by andre-merzky** [comment]
timeout is now implemented in devel-andre.  That will need a while to merge into devel though, as as a release of SAGA is imminent as we have frozen devel for non-hotfix merged.  In a week or two it should trickle down though...

In that branch, you can set the timeout via

```
export SAGA_PTY_SSH_TIMEOUT=6.5
```

**Comment by jcohen02** [comment]
Many thanks, however still experiencing a problem getting the timeout to work. I also note that debug output is not being shown in devel-andre - is there an alternative approach to turing on debug output?

I've tried setting the timeout both as an environment variable in the shell where I'm running the test code and using os.environ at the start of the test script. 

If I attempt to connect to an IP where there is no operational resource, the code hangs for a long time, even if I set SAGA_PTY_SSH_TIMEOUT to, say, 3.0. 

If I simply try to SSH to the resource, e.g. ssh x.x.x.x with no options, I see the same behaviour as with SAGA, the attempt to connect hangs waiting for the connection to be made. However, if I use ssh -o ConnectTimeout=3 x.x.x.x then the connection attempt fails with a "timed out" message after three seconds which is what I'm hoping for through setting SAGA_PTY_SSH_TIMEOUT.

Thanks. Jeremy.

**Comment by andre-merzky** [comment]
Ah, my apologies for the missing debug output: please set `RADICAL_SAGA_VERBOSE=DEBUG`.  I still need to add backward compatibility for `SAGA_VERBOSE` in that branch :/

re timeout: thanks, I'll recheck if the timeout indeed applies for that configuration.  It seemed to do so for me.  Hmmm....

**Comment by jcohen02** [comment]
I've pushed a suggested fix (jcohen02/saga-python@a2be801af1ccc209f898cecad13c46a3cd559928) to my fork which passes the value of `SAGA_PTY_SSH_TIMEOUT` to the `ConnectTimeout` option of the `ssh` command line.

If you think this looks reasonable, I'll create a pull request. I also merged in the commits from radical-cybertools/radical.utils#43 to make this easier to test.

**Comment by andre-merzky** [comment]
Thanks Jeremy, that indeed looks reasonable indeed!  A pull request would be appreciated.

**Comment by andre-merzky** [comment]
This has been merged now.  Thanks!
